File: clients/android/facetkit/src/main/kotlin/dev/fct/facetkit/StyleResolver.kt
package dev.fct.facetkit

/**
 * StyleResolver resolves a node's [Style] from its tag, classes, and inline
 * style — so live SSE fragments (parsed on-device) get the SAME styling the server
 * resolves for the initial tree.
 *
 * NOTE: [classStyles] mirrors `fa/style.go` (the single source of truth). The
 * long-term plan is to push neutral, already-styled trees over SSE so this table
 * lives only on the server; until then this kept-in-sync copy makes updates
 * pixel-consistent with the web.
 */
object StyleResolver {

    fun resolve(tag: String, attrs: Map<String, String>): Style? {
        var s = Style()
        if (tag == "button" || tag == "a") s = s.copy(direction = "row", align = "center")
        attrs["class"]?.split(Regex("\\s+"))?.forEach { c ->
            classStyles[c]?.let { s = merge(s, it) }
        }
        attrs["style"]?.let { s = applyInline(s, it) }
        return if (s == Style()) null else s
    }

    private fun merge(a: Style, b: Style) = Style(
        direction = b.direction ?: a.direction,
        gap = b.gap ?: a.gap,
        pad = b.pad ?: a.pad,
        align = b.align ?: a.align,
        justify = b.justify ?: a.justify,
        grow = b.grow ?: a.grow,
        width = b.width ?: a.width,
        bg = b.bg ?: a.bg,
        fg = b.fg ?: a.fg,
        fontSize = b.fontSize ?: a.fontSize,
        fontWeight = b.fontWeight ?: a.fontWeight,
        radius = b.radius ?: a.radius,
    )

    private fun applyInline(start: Style, inline: String): Style {
        var s = start
        for (decl in inline.split(";")) {
            val idx = decl.indexOf(':')
            if (idx < 0) continue
            val prop = decl.substring(0, idx).trim().lowercase()
            val v = decl.substring(idx + 1).trim()
            s = when (prop) {
                "width" -> s.copy(width = v)
                "background", "background-color" -> s.copy(bg = v)
                "color" -> s.copy(fg = v)
                "padding" -> s.copy(pad = px(v))
                "border-radius" -> s.copy(radius = px(v))
                "font-size" -> s.copy(fontSize = px(v))
                "font-weight" -> s.copy(fontWeight = v.toIntOrNull() ?: if (v == "bold") 700 else null)
                "gap" -> s.copy(gap = px(v))
                "flex-direction" -> if (v == "row" || v == "column") s.copy(direction = v) else s
                "display" -> if (v == "flex" && s.direction == null) s.copy(direction = "row") else s
                "justify-content" -> s.copy(justify = mapJustify(v))
                "align-items" -> s.copy(align = mapAlign(v))
                else -> s
            }
        }
        return s
    }

    private fun px(v: String): Int {
        var t = v.trim().removeSuffix("px")
        val dot = t.indexOf('.')
        if (dot >= 0) t = t.substring(0, dot)
        return t.trim().toIntOrNull() ?: 0
    }

    private fun mapJustify(v: String) = when (v) {
        "center" -> "center"; "flex-end", "end" -> "end"; "space-between" -> "between"; else -> "start"
    }

    private fun mapAlign(v: String) = when (v) {
        "center" -> "center"; "flex-end", "end" -> "end"; "stretch" -> "stretch"; else -> "start"
    }

    // Mirror of fa/style.go classStyles.
    private val classStyles: Map<String, Style> = mapOf(
        "fa-row" to Style(direction = "row", align = "center", gap = 8),
        "fa-post__header" to Style(direction = "row", gap = 10),
        "fa-post__actions" to Style(direction = "row", justify = "between"),
        "fa-vidctl" to Style(direction = "row", align = "center", gap = 10, pad = 8),
        "fa-engage" to Style(direction = "row", gap = 8),
        "fa-feedtabs" to Style(direction = "row"),
        "fa-tabs" to Style(direction = "row"),
        "fa-storybar" to Style(direction = "row", gap = 12, pad = 12),
        "fa-catchips" to Style(direction = "row", gap = 8),
        "fa-roomctl" to Style(direction = "row", align = "center", gap = 12, justify = "center"),
        "fa-composer" to Style(direction = "row", gap = 10, pad = 12),
        "fa-composer__bar" to Style(direction = "row", justify = "between", align = "center"),
        "fa-composer__tools" to Style(direction = "row"),
        "fa-setrow" to Style(direction = "row", justify = "between", align = "center", pad = 12),
        "fa-bottomnav" to Style(direction = "row", justify = "between", pad = 8),
        "fa-spacebar" to Style(direction = "row", align = "center", gap = 10, pad = 10),
        "fa-subrow" to Style(direction = "row", align = "center", gap = 10, pad = 10),
        "fa-sresult" to Style(direction = "row", align = "center", gap = 10, pad = 10),
        "fa-navrail__item" to Style(direction = "row", align = "center", gap = 14, pad = 10),
        "fa-roomhead" to Style(direction = "row", align = "center", gap = 12, pad = 12),
        "fa-topbar" to Style(direction = "row", align = "center", justify = "between", pad = 10),
        "fa-vcard__row" to Style(direction = "row", gap = 10),
        "fa-chatcompose" to Style(direction = "row", gap = 6, pad = 8),
        "fa-stack" to Style(direction = "column"),
        "fa-card" to Style(direction = "column", pad = 16, radius = 12, gap = 8),
        "fa-composer__main" to Style(direction = "column", gap = 8, grow = true),
        "fa-vcard__meta" to Style(direction = "column"),
        "fa-rrcard" to Style(direction = "column", pad = 12, radius = 16, gap = 8),
        "fa-btn" to Style(direction = "row", align = "center", pad = 8, radius = 999, fontWeight = 600),
        "fa-btn--primary" to Style(bg = "#1d9bf0", fg = "#ffffff"),
        "fa-btn--secondary" to Style(fg = "#0f1419"),
        "fa-btn--danger" to Style(bg = "#f4212e", fg = "#ffffff"),
        "fa-badge" to Style(pad = 4, radius = 999, fontSize = 12, fontWeight = 700),
        "fa-pill" to Style(pad = 6, radius = 999, fontSize = 12, fontWeight = 700),
        "fa-tip" to Style(direction = "row", align = "center", gap = 6, pad = 8, radius = 999, bg = "#ffc107", fontWeight = 800),
        "fa-post__name" to Style(fontWeight = 700),
        "fa-statcard__value" to Style(fontSize = 28, fontWeight = 800),
        "fa-channelhead__name" to Style(fontSize = 24, fontWeight = 800),
    )
}

File: clients/swift/Sources/FacetKit/StyleResolver.swift
import Foundation

/// StyleResolver resolves a node's `Style` from its tag, classes, and inline style,
/// so SSE fragments parsed on-device get the SAME styling the server resolves for
/// the initial tree.
///
/// NOTE: `classStyles` mirrors `fa/style.go` (the single source of truth). The
/// long-term plan is to push neutral, already-styled trees over SSE so this table
/// lives only on the server.
enum StyleResolver {

    static func resolve(tag: String, attrs: [String: String]) -> Style? {
        var s = Style()
        if tag == "button" || tag == "a" { s.direction = "row"; s.align = "center" }
        if let cls = attrs["class"] {
            for c in cls.split(separator: " ") {
                if let partial = classStyles[String(c)] { merge(&s, partial) }
            }
        }
        if let inline = attrs["style"] { applyInline(&s, inline) }
        return isZero(s) ? nil : s
    }

    private static func isZero(_ s: Style) -> Bool {
        s.direction == nil && s.gap == nil && s.pad == nil && s.align == nil &&
        s.justify == nil && s.grow == nil && s.width == nil && s.bg == nil &&
        s.fg == nil && s.fontSize == nil && s.fontWeight == nil && s.radius == nil
    }

    private static func merge(_ s: inout Style, _ o: Style) {
        if let v = o.direction { s.direction = v }
        if let v = o.gap { s.gap = v }
        if let v = o.pad { s.pad = v }
        if let v = o.align { s.align = v }
        if let v = o.justify { s.justify = v }
        if let v = o.grow { s.grow = v }
        if let v = o.width { s.width = v }
        if let v = o.bg { s.bg = v }
        if let v = o.fg { s.fg = v }
        if let v = o.fontSize { s.fontSize = v }
        if let v = o.fontWeight { s.fontWeight = v }
        if let v = o.radius { s.radius = v }
    }

    private static func applyInline(_ s: inout Style, _ inline: String) {
        for decl in inline.split(separator: ";") {
            guard let colon = decl.firstIndex(of: ":") else { continue }
            let prop = decl[..<colon].trimmingCharacters(in: .whitespaces).lowercased()
            let val = decl[decl.index(after: colon)...].trimmingCharacters(in: .whitespaces)
            switch prop {
            case "width": s.width = val
            case "background", "background-color": s.bg = val
            case "color": s.fg = val
            case "padding": s.pad = px(val)
            case "border-radius": s.radius = px(val)
            case "font-size": s.fontSize = px(val)
            case "font-weight": s.fontWeight = Int(val) ?? (val == "bold" ? 700 : nil)
            case "gap": s.gap = px(val)
            case "flex-direction": if val == "row" || val == "column" { s.direction = val }
            case "display": if val == "flex", s.direction == nil { s.direction = "row" }
            case "justify-content": s.justify = mapJustify(val)
            case "align-items": s.align = mapAlign(val)
            default: break
            }
        }
    }

    private static func px(_ v: String) -> Int {
        var t = v.hasSuffix("px") ? String(v.dropLast(2)) : v
        if let dot = t.firstIndex(of: ".") { t = String(t[..<dot]) }
        return Int(t.trimmingCharacters(in: .whitespaces)) ?? 0
    }

    private static func mapJustify(_ v: String) -> String {
        switch v { case "center": return "center"; case "flex-end", "end": return "end"
        case "space-between": return "between"; default: return "start" }
    }
    private static func mapAlign(_ v: String) -> String {
        switch v { case "center": return "center"; case "flex-end", "end": return "end"
        case "stretch": return "stretch"; default: return "start" }
    }

    // Mirror of fa/style.go classStyles.
    static let classStyles: [String: Style] = [
        "fa-row": Style(direction: "row", gap: 8, align: "center"),
        "fa-post__header": Style(direction: "row", gap: 10),
        "fa-post__actions": Style(direction: "row", justify: "between"),
        "fa-vidctl": Style(direction: "row", gap: 10, pad: 8, align: "center"),
        "fa-engage": Style(direction: "row", gap: 8),
        "fa-feedtabs": Style(direction: "row"),
        "fa-tabs": Style(direction: "row"),
        "fa-storybar": Style(direction: "row", gap: 12, pad: 12),
        "fa-catchips": Style(direction: "row", gap: 8),
        "fa-roomctl": Style(direction: "row", gap: 12, align: "center", justify: "center"),
        "fa-composer": Style(direction: "row", gap: 10, pad: 12),
        "fa-composer__bar": Style(direction: "row", align: "center", justify: "between"),
        "fa-composer__tools": Style(direction: "row"),
        "fa-setrow": Style(direction: "row", pad: 12, align: "center", justify: "between"),
        "fa-bottomnav": Style(direction: "row", pad: 8, justify: "between"),
        "fa-spacebar": Style(direction: "row", gap: 10, pad: 10, align: "center"),
        "fa-subrow": Style(direction: "row", gap: 10, pad: 10, align: "center"),
        "fa-sresult": Style(direction: "row", gap: 10, pad: 10, align: "center"),
        "fa-navrail__item": Style(direction: "row", gap: 14, pad: 10, align: "center"),
        "fa-roomhead": Style(direction: "row", gap: 12, pad: 12, align: "center"),
        "fa-topbar": Style(direction: "row", pad: 10, align: "center", justify: "between"),
        "fa-vcard__row": Style(direction: "row", gap: 10),
        "fa-chatcompose": Style(direction: "row", gap: 6, pad: 8),
        "fa-stack": Style(direction: "column"),
        "fa-card": Style(direction: "column", gap: 8, pad: 16, radius: 12),
        "fa-composer__main": Style(direction: "column", gap: 8, grow: true),
        "fa-vcard__meta": Style(direction: "column"),
        "fa-rrcard": Style(direction: "column", gap: 8, pad: 12, radius: 16),
        "fa-btn": Style(direction: "row", pad: 8, align: "center", fontWeight: 600, radius: 999),
        "fa-btn--primary": Style(bg: "#1d9bf0", fg: "#ffffff"),
        "fa-btn--secondary": Style(fg: "#0f1419"),
        "fa-btn--danger": Style(bg: "#f4212e", fg: "#ffffff"),
        "fa-badge": Style(pad: 4, fontSize: 12, fontWeight: 700, radius: 999),
        "fa-pill": Style(pad: 6, fontSize: 12, fontWeight: 700, radius: 999),
        "fa-tip": Style(direction: "row", gap: 6, pad: 8, align: "center", bg: "#ffc107", fontWeight: 800, radius: 999),
        "fa-post__name": Style(fontWeight: 700),
        "fa-statcard__value": Style(fontSize: 28, fontWeight: 800),
        "fa-channelhead__name": Style(fontSize: 24, fontWeight: 800),
    ]
}
